### ISSUE TITLE
smafa encountered the DeserializeBadBool error when trying to decode the specified database file

### ISSUE THREAD

**Issue by yinjunshiliu** [wwood/singlem#178]
Hello, I am currently using singlem to annotate my metagenomic data. I installed the conda version (v0.16.0), downloaded the metapackages (S4.2.2.GTDB_r214.metapackage_20240502.smpkg.zb, {"singlem_database_version": 5}) following the installation guide.

This is my command:
singlem pipe -1 A_1.fq.gz -2 A_2.fq.gz -p test.txt --metapackage /My_PATH/S4.2.2.GTDB_r214.metapackage_20240502.smpkg.zb -threads 4
and I received the following message:
05/07/2024 05:27:49 AM INFO: SingleM v0.16.0
05/07/2024 05:27:49 AM INFO: Acquiring SingleM metapackage from Zenodo backpack directory specified ..
05/07/2024 05:27:49 AM INFO: Retrieval successful. Location of backpack is: /My_PATH/S4.2.2.GTDB_r214.metapackage_20240502.smpkg.zb
05/07/2024 05:27:49 AM INFO: Loaded 59 SingleM packages
05/07/2024 05:27:50 AM INFO: Using as input 1 different pairs of sequence files e.g. A_1.fq.gz & A_2.fq.gz
05/07/2024 05:27:50 AM INFO: Filtering sequence files through DIAMOND blastx
05/07/2024 07:08:25 AM INFO: Finished DIAMOND prefilter phase
05/07/2024 07:08:25 AM INFO: Assigning sequences to SingleM packages with DIAMOND ..
05/07/2024 07:14:32 AM INFO: Running taxonomic assignment ..
05/07/2024 07:14:32 AM INFO: Assigning taxonomy by singlem query ..
05/07/2024 07:14:33 AM INFO: Querying against species database with 603 sequences, using method smafa-naive and max divergence 2
05/07/2024 07:14:33 AM INFO: Searching with SMAFA NAIVE by nucleotide sequence ..
05/07/2024 07:14:33 AM INFO: Querying index for S3.1.ribosomal_protein_L2_rplB
Traceback (most recent call last):
  File "***/miniconda3/envs/singlem/bin/singlem", line 658, in <module>
    singlem.pipe.SearchPipe().run(
  File "***/miniconda3/envs/singlem/lib/python3.10/site-packages/singlem/pipe.py", line 69, in run
    otu_table_object = self.run_to_otu_table(**kwargs)
  File "***/miniconda3/envs/singlem/lib/python3.10/site-packages/singlem/pipe.py", line 456, in run_to_otu_table
    otu_table_object = self.assign_taxonomy_and_process(
  File "***/miniconda3/envs/singlem/lib/python3.10/site-packages/singlem/pipe.py", line 491, in assign_taxonomy_and_process
    assignment_result = self._assign_taxonomy(
  File "***/miniconda3/envs/singlem/lib/python3.10/site-packages/singlem/pipe.py", line 1212, in _assign_taxonomy
    query_based_assignment_result = PipeTaxonomyAssignerByQuery().assign_taxonomy(
  File "***/miniconda3/envs/singlem/lib/python3.10/site-packages/singlem/pipe_taxonomy_assigner_by_query.py", line 120, in assign_taxonomy
    query_single_set(queries[0], 0)
  File "***/miniconda3/envs/singlem/lib/python3.10/site-packages/singlem/pipe_taxonomy_assigner_by_query.py", line 103, in query_single_set
    for hit in querier.query_with_queries(queries, sdb, max_species_divergence, method, SequenceDatabase.NUCLEOTIDE_TYPE, 1, None, False, None):
  File "***/miniconda3/envs/singlem/lib/python3.10/site-packages/singlem/querier.py", line 395, in query_by_sequence_similarity_with_smafa_naive
    smafa_stdout = extern.run(smafa_cmd, stdin='\n'.join([">{}\n{}".format(i, q.sequence) for i, q in enumerate(chunked_queries)]))
  File "***/miniconda3/envs/singlem/lib/python3.10/site-packages/extern/__init__.py", line 41, in run
    raise ExternCalledProcessError(process, command)
extern.ExternCalledProcessError: Command smafa query --database '/My_PATH/S4.2.2.GTDB_r214.metapackage_20240502.smpkg.zb/payload_directory/transcripts.sdb/nucleotide_indices_smafa_naive/S3.1.ribosomal_protein_L2_rplB.smafa_naive_index' --query /dev/stdin  --max-divergence 2 --max-num-hits 1 returned non-zero exit status 1.
STDERR was: b'[2024-05-07T11:14:33Z INFO  bird_tool_utils::clap_utils] Smafa version 0.7.1\n[2024-05-07T11:14:33Z INFO  smafa] Decoding db file /My_PATH/S4.2.2.GTDB_r214.metapackage_20240502.smpkg.zb/payload_directory/transcripts.sdb/nucleotide_indices_smafa_naive/S3.1.ribosomal_protein_L2_rplB.smafa_naive_index\nError: DeserializeBadBool\n'STDOUT was: b''

I wonder what I can do to avoid this error, thanks.

**Comment by wwood** [comment]
Hi,

Sorry for this mess - it is because you are using a version of smafa which is too old for the DB, because we are in the middle of a (messy) release of the whole stack.

Probably the easiest solution would be to manually get the old metapackage, by downloading (and un-tar-gzipping) from https://zenodo.org/records/8419620 and using that instead of the one that singlem data grabs.

Or you could just wait for the new release, which shouldn't be more than a week away.

**Comment by yinjunshiliu** [comment]
Thank you for your reply,

I downloaded the the old metapackage as you provided and rerun singlem with the following command:

singlem pipe -1 A_1.fq.gz -2 A_2.fq.gz -p test_out.txt --metapackage My_PATH/S3.2.1. GTDB_r214. Metapackage_20231006. SMPKG. Zb

The result file has been successfully obtained, and the run log is as follows:

05/11/2024 10:03:00 AM INFO: SingleM v0.16.0
05/11/2024 10:03:00 AM INFO: Acquiring SingleM metapackage from Zenodo backpack directory specified ..
05/11/2024 10:03:00 AM INFO: Retrieval successful. Location of backpack is: My_PATH/S3.2.1.GTDB_r214.metapackage_20231006.smpkg.zb
05/11/2024 10:03:00 AM INFO: Loaded 59 SingleM packages
...
05/11/2024 12:05:54 PM INFO: Assigning taxonomy with DIAMOND for 624 and 646 out of 908 and 880 sequences (68.7% and 73.4%) for sample F-0.1_1, package S3.59.uS11_bact.spkg
05/11/2024 12:06:38 PM INFO: Finished running taxonomic assignment
05/11/2024 12:06:59 PM WARNING: Unexpected lack (or >1 found) of assignment method for read name LH00330:52:222KCNLT4:1:1153:21651:8676
05/11/2024 12:06:59 PM WARNING: Unexpected lack (or >1 found) of assignment method for read name LH00330:52:222KCNLT4:1:2120:43824:27979
05/11/2024 12:06:59 PM WARNING: Unexpected lack (or >1 found) of assignment method for read name LH00330:52:222KCNLT4:1:2120:43816:27993
05/11/2024 12:06:59 PM WARNING: Unexpected lack (or >1 found) of assignment method for read name LH00330:52:222KCNLT4:1:2120:43808:27979
05/11/2024 12:06:59 PM WARNING: Unexpected lack (or >1 found) of assignment method for read name LH00330:52:222KCNLT4:1:1170:6163:6603
...
05/11/2024 12:07:35 PM INFO: Finished
05/11/2024 12:07:35 PM INFO: Writing taxonomic profile to test.txt
05/11/2024 12:07:35 PM INFO: Using minimum taxon coverage of 0.35
05/11/2024 12:07:35 PM INFO: Removing off-target OTUs from A_1
05/11/2024 12:07:35 PM INFO: Found 130858.41 assigned and 88.48 unassigned OTU coverage units
05/11/2024 12:07:35 PM INFO: After removing off-target OTUs, found 112493.14 assigned and 0.00 unassigned OTU coverage units
05/11/2024 12:07:35 PM INFO: Total OTU coverage by query: 20969.08552538581
05/11/2024 12:07:35 PM INFO: Total OTU coverage by diamond: 91524.05181946977
05/11/2024 12:07:35 PM INFO: Applying species-wise expectation maximization algorithm to OTU table
05/11/2024 12:07:35 PM INFO: Found 14 species uniquely hitting >= 10 marker genes
05/11/2024 12:07:49 PM INFO: Species-wise EM converged in 56 steps
05/11/2024 12:07:49 PM INFO: Gathering equivalence classes
05/11/2024 12:07:49 PM INFO: Demultiplexing OTU table
05/11/2024 12:07:50 PM INFO: Finished expectation maximization
05/11/2024 12:07:50 PM INFO: Converting DIAMOND IDs to taxons
05/11/2024 12:07:51 PM INFO: Converted 33443 Diamond-assigned OTU taxon_ids to taxon strings
05/11/2024 12:07:51 PM INFO: Applying genus-wise expectation maximization algorithm to OTU table
05/11/2024 12:08:01 PM INFO: Genus-wise EM converged in 43 steps
05/11/2024 12:08:01 PM INFO: Gathering equivalence classes
05/11/2024 12:08:01 PM INFO: Demultiplexing OTU table
05/11/2024 12:08:01 PM INFO: Finished genus expectation maximization
05/11/2024 12:08:02 PM INFO: Total profile coverage after condense domain to species: 3107.267659407961
05/11/2024 12:08:02 PM INFO: Total profile coverage after push down: 3107.267659407965
05/11/2024 12:08:02 PM INFO: Taxonomic level coverage:
05/11/2024 12:08:02 PM INFO: kingdom:   1.30%   2 taxons
05/11/2024 12:08:02 PM INFO: phylum:    1.51%   25 taxons
05/11/2024 12:08:02 PM INFO: class:     3.99%   61 taxons
05/11/2024 12:08:02 PM INFO: order:     6.09%   114 taxons
05/11/2024 12:08:02 PM INFO: family:    22.48%  257 taxons
05/11/2024 12:08:02 PM INFO: genus:     59.51%  643 taxons
05/11/2024 12:08:02 PM INFO: species:   5.11%   60 taxons
05/11/2024 12:08:02 PM INFO: Finished condense

In the run log, there are several warning messages. I am wondering whether these warnings have no effect on the running results, and if so, whether they can be ignored. Also, I am looking forward to the new release and will try it out when it is available. Thanks again.
